### ISSUE TITLE
warn redundant casts doesn't work with `Protocol.__call__`

### ISSUE THREAD

**Issue by graingert** [python/mypy#12328]
**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.10&flags=strict%2Cwarn-redundant-casts&gist=bc9999020b0fb32610c0c3ec99a8568c

```python
from typing import Protocol, cast

class FooType(Protocol):
    def __call__(self, ham: str) -> str:
        pass

def foo(ham: str) -> str:
    pass


v = cast(FooType, foo)
u: FooType = foo
```

**Expected Behavior**
I was expecting `v = cast(FooType, foo)` to work about the same as `v: FooType = foo # type: ignore[assignment]` with warn redundant ignores

**Your Environment**

See mypy-play

**Comment by jhance** [comment]
Is the behavior with respect to subtyping documented?

**Comment by graingert** [comment]
The docs refer to casts that can be safely removed:

https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-redundant-casts

> This flag will make mypy report an error whenever your code uses an unnecessary cast that can safely be removed.
